Update the teleport-kube-agent reference

ptgott opened the pull request:
Fixes #17578

Document values fields and behavior introduced by #13942.

Since the teleport-kube-agent chart now deploys more resources--and deploys them based on more sophisticated logic--this change lists the resources deployed by this chart and describes the conditions in which this chart deploys them.

ptgott committed:
Update the teleport-kube-agent reference

Fixes #17578

Document values fields and behavior introduced by #13942.

Since the teleport-kube-agent chart now deploys more resources--and
deploys them based on more sophisticated logic--this change lists the
resources deployed by this chart and describes the conditions in which
this chart deploys them.

tigrato commented on docs/pages/reference/helm-reference/teleport-kube-agent.mdx:
```diff
@@ -17,9 +34,29 @@ The `teleport-kube-agent` chart can run any or all of three Teleport services:
+|`ServiceAccount`|`serviceAccountName`, if given, or the release name|Enables the Teleport pod to manage access to resources in the Kubernetes cluster.|`serviceAccount.create` is `true`|
+|`PodDisruptionBudget`|The release name|Ensure high availability for the Teleport pod.|`highAvailability.podDisruptionBudget.enabled` is `true`.|
+|`ServiceAccount`|The release name, suffixed by `-hook`|Used to delete legacy `Deployment`s in order to deploy a `StatefulSet` instead.|If the `teleport-kube-agent` release contains a legacy `Deployment` resource.|
+|`Role`|The release name, suffixed by `-hook`|Used to delete legacy `Deployment`s in order to deploy a `StatefulSet` instead.|If the `teleport-kube-agent` release contains a legacy `Deployment` resource.|
```
We should mention that every `*-hook` resources are removed once the upgrade finishes

tigrato commented on docs/pages/reference/helm-reference/teleport-kube-agent.mdx:
```diff
@@ -59,6 +96,90 @@ This parameter is not mandatory to preserve backwards compatibility with older c
+  </TabItem>
+</Tabs>
+
+## `serviceAccountName`
```
This field is deprecated. Instead, everyone should use `serviceAccount.name`

tigrato commented on docs/pages/reference/helm-reference/teleport-kube-agent.mdx:
```diff
@@ -804,6 +925,31 @@ Enables the creation of a Kubernetes persistent volume to hold Teleport instance
 
+#### Chart upgrades
+
+If `storage.enabled` is `false` and you are upgrading a legacy
+`teleport-kube-agent` chart, the chart will launch a Teleport pod as a
+`Deployment` instead of a `StatefulSet`, disabling persistent storage for the
+Teleport pod. If `storage.enabled` is set to `true` during an upgrade, the chart
+will create a webhook that deletes any legacy `Deployment` and launches a
+`StatefulSet` instead.
```
Depending on the storage value, the upgrade process is different:

# `storage.enabled=false`

Previously, the Teleport pod was part of a `Deployment` resource.
During the upgrade, a `Statefulset` resource is installed alongside the `Deployment`.
Once the Statefulset joins the cluster and becomes healthy, the chart deletes the `Deployment` object using a Helm Hook, and the upgrade finishes.

# `storage.enabled=true`
Previously, the Teleport pod was part of a `StatefulSet` resource.
During the upgrade, the resource is updated (kept as `StatefulSet`), and the identity is imported from external storage into the Kubernetes secret.

tigrato commented on docs/pages/reference/helm-reference/teleport-kube-agent.mdx:
```diff
@@ -804,6 +925,31 @@ Enables the creation of a Kubernetes persistent volume to hold Teleport instance
+#### New chart installations
+
+New installations of `teleport-kube-agent` always launch the Teleport pod as a
+`StatefulSet`, and `storage.enabled` configures persistent storage for the pod.
```
After T11, Storage is no longer used to store Teleport Identity even if storage is enabled.
Storage is useful if you want to store session recordings locally.

ptgott committed:
Respond to PR feedback

- Fix the legacy behavior description at the top of the refernece. Also
  add this to its own H3 since it is getting a bit lengthy.

- Note that *-hook resources are removed after the upgrade finishes.

- Note that `serviceAccountName` is deprecated and expand the
  `serviceAccount.name` entry a bit.

- Add a note re: storage not being required to the "New chart
  installations" section.

- Clarify `storage.enabled` in relation to upgrades

Commit:
format tables

Commit:
add br tags to service table

In order to display the service name without a line break mid-word.

Commit:
adjust br tag placement

To avoid breaking the first column

alexfornuto commented on docs/pages/reference/helm-reference/teleport-kube-agent.mdx:
```diff
@@ -3,39 +3,86 @@ title: teleport-kube-agent Chart Reference
+
+The `teleport-kube-agent` chart deploys the following Kubernetes resources:
+
+| Kind                  | Default Name                                               | Description                                                                                                                                                                                                                                                                                             | When Deployed  ...
+|-----------------------|------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------- ...
+| `StatefulSet`         | The release name                                           | Running a user-configured Teleport pod.                                                                                                                                                                                                                                                                 | Always.        ...
+| `Secret`              | `secretName` (default: `teleport-kube-agent-join-token`)   | Used for managing the state of the Teleport pod.                                                                                                                                                                                                                                                        | `authToken` or ...
+| `Deployment`          | The release name                                           | Runs a user-configured Teleport pod.                                                                                                                                                                                                                                                                    | `storage.enabl ...
+| `Role`                | The `roleName` option, if given, or the release name.      | Used to manage the state of the Teleport pod via Kubernetes secrets.                                                                                                                                                                                                                                    | Always.        ...
+| `ClusterRole`         | `clusterRoleName`, if given, or the release name.          | Allows impersonating users, groups, and service accounts, getting pods, and creating [`SelfSubjectAccessReview`s](https://www.pulumi.com/registry/packages/kubernetes/api-docs/authorization/v1/selfsubjectaccessreview/) so the Teleport pod can manage access to resources in its Kubernetes cluster. | Always.        ...
+| `ClusterRoleBinding`  | `clusterRoleBindingName`, if provided, or the release name | Enables the Teleport pod to manage access to resources in the Kubernetes cluster.                                                                                                                                                                                                                       | Always.        ...
+| `RoleBinding`         | `roleBindingName`, if given, or the release name           | Enables the Teleport pod to manage access to resources in the Kubernetes cluster.                                                                                                                                                                                                                       | Always.        ...
+| `ServiceAccount`      | `serviceAccount.name`, if given, or the release name       | Enables the Teleport pod to manage access to resources in the Kubernetes cluster.                                                                                                                                                                                                                       | `serviceAccoun ...
+| `PodDisruptionBudget` | The release name                                           | Ensure high availability for the Teleport pod.                                                                                                                                                                                                                                                          | `highAvailabil ...
+| `ServiceAccount`      | The release name, suffixed by `-hook`                      | Used to delete legacy `Deployment`s in order to deploy a `StatefulSet` instead. Removed once the upgrade is complete.                                                                                                                                                                                   | If the `telepo ...
+| `Role`                | The release name, suffixed by `-hook`                      | Used to delete legacy `Deployment`s in order to deploy a `StatefulSet` instead. Removed once the upgrade is complete.                                                                                                                                                                                   | If the `telepo ...
+| `RoleBinding`         | The release name, suffixed by `-hook`                      | Used to delete legacy `Deployment`s in order to deploy a `StatefulSet` instead. Removed once the upgrade is complete.                                                                                                                                                                                   | If the `telepo ...
+| `Job`                 | The release name, suffixed by `-hook`                      | Used to delete legacy `Deployment`s in order to deploy a `StatefulSet` instead. Removed once the upgrade is complete.                                                                                                                                                                                   | If the `telepo ...
+| `ConfigMap`           | The release name                                           | Contains the configuration for the Teleport pod.                                                                                                                                                                                                                                                        | Always.        ...
+| `PodSecurityPolicy`   | The release name                                           | Enforces security requirements for pods deployed by `teleport-kube-agent`.                                                                                                                                                                                                                              | `podSecurityPo ...
+| `Role`                | The release name, suffixed by `-psp`                       | Enforces security requirements for pods deployed by `teleport-kube-agent`.                                                                                                                                                                                                                              | `podSecurityPo ...
+| `RoleBinding`         | The release name, suffixed by `-psp`                       | Enforces security requirements for pods deployed by `teleport-kube-agent`.                                                                                                                                                                                                                              | `podSecurityPo ...
```
I tried to apply `<br/>` tags to this table so the fist column wouldn't break, but couldn't find the magic placements to make it work. IMHO this is another example of why we should prioritize https://github.com/gravitational/docs/issues/126.

alexfornuto commented on docs/pages/reference/helm-reference/teleport-kube-agent.mdx:
```diff
@@ -59,11 +106,72 @@ This parameter is not mandatory to preserve backwards compatibility with older c
+## `serviceAccountName`
+
+This field is deprecated and will be removed in a future version. Use
+`serviceAccount.name` instead.
```
```suggestion
[`serviceAccount.name`](#serviceaccountname-1) instead.
```

alexfornuto reviewed (approved):
In addition to my comments below I pushed a few commits. I intentionally made more small commits so you could review the changes more easily. Feel free to squash or drop them as you see fit.

👍 pending outstanding comments.

ptgott commented on docs/pages/reference/helm-reference/teleport-kube-agent.mdx:
```diff
@@ -3,11 +3,20 @@ title: teleport-kube-agent Chart Reference
-The `teleport-kube-agent` Helm chart is used to configure a Teleport instance
-that runs in a remote Kubernetes cluster and joins back to a Teleport cluster to
-provide access to services running there.
+The `teleport-kube-agent` Helm chart is used to configure a Teleport agent that
+runs in a remote Kubernetes cluster and joins to a Teleport cluster to provide
+access to resources in your infrastructure. 
```
I've simplified this/made it more general.

ptgott committed:
Respond to PR feedback
